dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java

```diff
@@ -124,12 +130,20 @@ public static Init newInit(MemoryObject base, int offset) {
         return new Init(base, offset, address);
     }
 
-    public static DirectValueFunctionCall newValueFunctionCall(Register resultRegister, Function function, List<Expression> arguments) {
-        return new DirectValueFunctionCall(resultRegister, function, arguments);
+    public static ValueFunctionCall newValueFunctionCall(Register resultRegister, Function function, List<Expression> arguments) {
+        return new ValueFunctionCall(resultRegister, function.getFunctionType(), function, arguments);
     }
 
-    public static DirectVoidFunctionCall newVoidFunctionCall(Function function, List<Expression> arguments) {
-        return new DirectVoidFunctionCall(function, arguments);
+    public static ValueFunctionCall newValueFunctionCall(Register resultRegister, FunctionType funcType,
+                                                         Expression funcPtr, List<Expression> arguments) {
+        return new ValueFunctionCall(resultRegister, funcType, funcPtr, arguments);
+    }
+
+    public static VoidFunctionCall newVoidFunctionCall(Function function, List<Expression> arguments) {
+        return new VoidFunctionCall(function.getFunctionType(), function, arguments);
```

Comment:
Does this guarantee that the return type is actually `void`?

Reply:
I think the constructor of `VoidFunctionCall` checks this.